src/System.Text.Encodings.Web.Utf8/UrlEncoder.cs

```diff
@@ -251,80 +251,29 @@ private static int UnescapePercentEncoding(ref int scan, Span<byte> buffer)
             {
                 return -1;
             }
-
-            var probe = scan;
-
-            var value1 = ReadHex(ref probe, buffer);
-            if (value1 == -1)
+            
+            if (scan + 2 > buffer.Length)
             {
                 return -1;
             }
+            
+            byte value;
+            int bytesConsumed;
+            bool result = PrimitiveParser.InvariantUtf8.Hex.TryParseByte(buffer.Slice(scan, 2), out value, out bytesConsumed);
 
-            var value2 = ReadHex(ref probe, buffer);
-            if (value2 == -1)
-            {
-                return -1;
-            }
-
-            if (SkipUnescape(value1, value2))
-            {
-                return -1;
-            }
-
-            scan = probe;
-            return (value1 << 4) + value2;
-        }
-
-
-        /// <summary>
-        /// Read the next char and convert it into hexadecimal value.
-        /// 
-        /// The <paramref name="scan"/> index will be moved to the next
-        /// byte no matter no matter whether the operation successes.
-        /// </summary>
-        /// <param name="scan">The index of the byte in the buffer to read</param>
-        /// <param name="buffer">The byte span from which the hex to be read</param>
-        /// <returns>The hexadecimal value if successes, otherwise -1.</returns>
-        private static int ReadHex(ref int scan, Span<byte> buffer)
-        {
-            if (scan == buffer.Length)
+            if (result == false || bytesConsumed != 2)
```

Comment:
OK. I just haven't seen this style used anywhere and I would imagine it is more prone to errors like result = false instead of result == false. I am not sure if there are specific guidelines around this. I would be curious if one exists.

haha: "...but this question will likely solicit debate, arguments, polling, or extended discussion."
http://stackoverflow.com/questions/7535818/using-if-bool-vs-if-bool-false-in-c-sharp